File: script-files/measures/outputsContinuous/generalized_logistic.js
jStat = require("../node_modules/jstat");

function digamma(x) {
    return Math.log(x) - 1 / (2 * x);
}

function polygamma(k, x) {
    function suma(x) {
        let suma = 0;
        for (let i = 0; i < 1000; i++) {
            suma += 1 / Math.pow(x + i, k + 1);
        }
        return suma;
    }
    var resultado = Math.pow(-1, k + 1) * jStat.factorial(k) * suma(x);
    return resultado;
}

dists = {
    generalized_logistic: {
        measurements: {
            nonCentralMoments: function (k, c, loc, scale) {
                return undefined;
            },
            centralMoments: function (k, c, loc, scale) {
                return undefined;
            },
            stats: {
                mean: function (c, loc, scale) {
                    return loc + scale * (0.57721 + digamma(c));
                },
                variance: function (c, loc, scale) {
                    return scale * scale * (Math.PI * Math.PI / 6 + polygamma(1, c));
                },
                standardDeviation: function (c, loc, scale) {
                    return Math.sqrt(this.variance(c, loc, scale));
                },
                skewness: function (c, loc, scale) {
                    return (2.40411380631918 + polygamma(2, c)) / ((Math.PI * Math.PI / 6 + polygamma(1, c)) ** 1.5);
                },
                kurtosis: function (c, loc, scale) {
                    return 3 + (6.49393940226682 + polygamma(3, c)) / ((Math.PI * Math.PI / 6 + polygamma(1, c)) ** 2);
                },
                median: function (c, loc, scale) {
                    return dists.generalized_logistic.measurements.ppf(0.5, c, loc, scale);
                },
                mode: function (c, loc, scale) {
                    return loc + scale * Math.log(c);
                },
            },
        }
    }
}
console.log(dists.generalized_logistic.measurements.stats.mean(5, 10, 34))
console.log(dists.generalized_logistic.measurements.stats.variance(5, 10, 34))
console.log(dists.generalized_logistic.measurements.stats.standardDeviation(5, 10, 34))
console.log(dists.generalized_logistic.measurements.stats.skewness(5, 10, 34))
console.log(dists.generalized_logistic.measurements.stats.kurtosis(5, 10, 34))
// console.log(dists.generalized_logistic.measurements.stats.median(5, 10, 34))
console.log(dists.generalized_logistic.measurements.stats.mode(5, 10, 34))

// mean_value: 80.9460290227594
// variance_value: 2156.242296899568
// standard_deviation_value: 46.43535610824545
// skewness_value: 0.9245731210131928
// kurtosis_value: 4.872662204876139
// median_value: 74.79840659673239
// mode_value: 64.72088902275941

File: continuous/distributions/generalized_logistic.py
import scipy.stats
import scipy.special as sc
import math
import scipy.optimize
import numpy


class GENERALIZED_LOGISTIC:
    """
    Generalized Logistic Distribution
    Compendium of Common Probability Distributions (pag.41) ... Michael P. McLaughlin
    https://docs.scipy.org / doc / scipy / tutorial / stats / continuous_genlogistic.html
    """

    def __init__(self, measurements):
        self.parameters = self.get_parameters(measurements)

        self.loc = self.parameters["loc"]
        self.scale = self.parameters["scale"]
        self.c = self.parameters["c"]

    def cdf(self, x: float) -> float:
        """
        Cumulative distribution function
        Calculated using the definition of the function
        Alternative: quadrature integration method
        """
        # return scipy.stats.genlogistic.cdf(x, self.c, loc=self.loc, scale=self.scale)
        z = lambda t: (t - self.loc) / self.scale
        return 1 / ((1 + math.exp(-z(x))) ** self.c)

    def pdf(self, x: float) -> float:
        """
        Probability density function
        Calculated using definition of the function in the documentation
        """
        # return scipy.stats.genlogistic.pdf(x, self.c, loc=self.loc, scale=self.scale)
        z = lambda t: (t - self.loc) / self.scale
        return (self.c / self.scale) * math.exp(-z(x)) * ((1 + math.exp(-z(x))) ** (-self.c - 1))

    def get_num_parameters(self) -> int:
        """
        Number of parameters of the distribution
        """
        return len(self.parameters)

    def parameter_restrictions(self) -> bool:
        """
        Check parameters restrictions
        """
        v1 = self.scale > 0
        v2 = self.c > 0
        return v1 and v2

    def get_parameters(self, measurements) -> dict[str, float | int]:
        """
        Calculate proper parameters of the distribution from sample measurements.
        The parameters are calculated by formula.

        Parameters
        ==========
        measurements: MEASUREMESTS
            attributes: mean, std, variance, skewness, kurtosis, median, mode, min, max, length, num_bins, data

        Returns
        =======
        parameters : dict
            {"loc": * , "scale": * , "c": * }
        """

        def equations(initial_solution: tuple[float], measurements) -> tuple[float]:
            ## Variables declaration
            c, loc, scale = initial_solution

            ## Parametric expected expressions
            parametric_mean = loc + scale * (0.57721 + sc.digamma(c))
            parametric_variance = scale**2 * (math.pi**2 / 6 + sc.polygamma(1, c))
            # parametric_skewness = (sc.polygamma(2,1) + sc.polygamma(2,c)) / ((math.pi ** 2 / 6 + sc.polygamma(1, c)) ** 1.5)
            # parametric_kurtosis = 3 + (math.pi ** 4 / 15 + sc.polygamma(3,c)) / ((math.pi ** 2 / 6 + sc.polygamma(1, c)) ** 2)
            parametric_median = loc + scale * (-math.log(0.5 ** (-1 / c) - 1))
            # parametric_mode = loc + scale * math.log(c)

            ## System Equations
            eq1 = parametric_mean - measurements.mean
            eq2 = parametric_variance - measurements.variance
            # eq3 = parametric_skewness - measurements.skewness
            # eq3 = parametric_kurtosis - measurements.kurtosis
            eq3 = parametric_median - measurements.median
            # eq3 = parametric_mode - measurements.mode

            return (eq1, eq2, eq3)

        # ## scipy.optimize.fsolve methods
        # solution = scipy.optimize.fsolve(equations, (1, 1, 1), measurements)
        # parameters = {"loc": solution[0], "scale": solution[1], "c": solution[2]}
        # print(parameters)

        ## least square methods
        x0 = [measurements.mean, measurements.mean, measurements.mean]
        b = ((1e-5, -numpy.inf, 1e-5), (numpy.inf, numpy.inf, numpy.inf))
        solution = scipy.optimize.least_squares(equations, x0, bounds=b, args=([measurements]))
        parameters = {"c": solution.x[0], "loc": solution.x[1], "scale": solution.x[2]}

        # ## scipy methods
        # scipy_params = scipy.stats.genlogistic.fit(measurements.data)
        # parameters = {"loc": scipy_params[1], "scale": scipy_params[2], "c": scipy_params[0]}

        return parameters


if __name__ == "__main__":
    ## Import function to get measurements
    import sys

    sys.path.append("../measurements")
    from measurements_continuous import MEASUREMENTS_CONTINUOUS

    ## Import function to get measurements
    def get_data(direction):
        sample_distribution_file = open(direction, "r")
        data = [float(x.replace(",", ".")) for x in sample_distribution_file.read().splitlines()]
        return data

    ## Distribution class
    path = "../data/data_generalized_logistic.txt"
    data = get_data(path)
    measurements = MEASUREMENTS_CONTINUOUS(data)
    distribution = GENERALIZED_LOGISTIC(measurements)

    print(distribution.get_parameters(measurements))
    print(distribution.cdf(measurements.mean))
    print(distribution.pdf(measurements.mean))
